### DISCUSSION TITLE
Library sharing and people matching

### DISCUSSION THREAD

**Discussion by andymarden** [immich-app/immich#7362] [Q&A]
My setup is this: I have a "family" user which owns an external library which is where all of our family's historical photos are.
This is then shared with each family member as a user.
It's great that now th9is can show up on maps but my question is on people face matching. Unless I am missing something, each user gets to identify people in their own libraries so, in this case, the "family" user can go through and do that.
However, it seems this matching is not available to the users that the library is shared with. Of course, I would want all family members to be able to see this.

SO making that available would be good - even better would be if you could say "this person in the shared library is the same as this person in my own library" and link the people

Have I missed something or is this a new/existing enhancement request?

I could have every user create their own external library on the same folder but that is a HUGE increase in resources required and it means everyone doing all this stuff for themselves.

This must generally be a very common way to want to use Immich and I hope that this is a lens that is being considered as work progresses - each user really would want to see their own + the shared libraries in one view.

**Comment by andymarden** [comment]
I can see that looking further, the shared libraries are very separate in the users display which is a real shame

Maybe I need to rethink this at this stage and have everyone log on as the same user since the primary purpose is around sharing family photos - the ability to upload and store your own photos is 2nd to this.

But, as I said - it would be great if we could move to a more integrated experience.

**Comment by mikesellt** [comment]
I'm just going to add my thoughts here as well.  I'm moving from Synology to TrueNAS, and as a result am moving my photo library over to Immich.  I have a lot of photos already, and I'd like to be able to add them as an external library, but give all of my users (my wife and kids) view access to that external library while still keeping their own individual libraries for cell phone backup.  I don't see a way to do this currently without making individual external libraries for each user and adding the same photo folder in each of them - basically duplicating the external library for each user.  If that library were duplicated, I don't think all of the face recognition would persist between each library.  If we could just have one external library and have it shared between users, they could help match the faces and whatnot.  

Also, great project so far.  I'm really liking Immich in general.  Thanks.

**Reply by tuupola** [reply]
Yep, this is how it should work.

**Reply by mikesellt** [reply]
Hello tuupola,  I'm not sure what you mean by your reply.  Should facial recognition persist across shared libraries?  Or do you mean that I need to create an external library for each user that point to the same master folder?

**Reply by andymarden** [reply]
I hope the former in that all users (or a singer with permissions) can not only view all things related to the library but also can contribute to reviewing and updating facial recognition, etc, such then everyone can see the same thing.

**Comment by rytl4847** [comment]
I also hope that this gets implemented in the future. I have the exact same use case as @andymarden.

The option to include shared library resources in the map view was added last year after a similar request.

**Comment by optoisolated** [comment]
It should be possible as in when you share a library with a partner you can share the location information, you should also be able to share the people as well. Even if its a selectable option. 

Our library is set up in a similar way with two individual users but we would like to see each others libraries and having to people-match on each library separately is laborious.

**Comment by andymarden** [comment]
Ultimately this is about the underlying data model. At the moment, library is something owned by user and tied to it.

It would be better to think of them as separate entities/classes with a nanny to many relationship between them. 

The link entity between them drives the users and the access they have to the libraries eg. view photos, view locations, view face recognition, etc and then edit for each of these things on a per library basis.

There has already been a recognition of this with external libraries moving out from user ownership, but take it to its logical conclusion send all the flexibility required with be there.

There's a load of work to do that but it would be worth it in the long run.

**Reply by Jacup** [reply]
library could be owned by 1 person to manage that, but maybe add additional field with collection of users eligible to view contents? Maybe even a little permissions management with roles like viewer, view+modify etc.

**Comment by Jacup** [comment]
As an WA, what about creating user, that will own an external library and share it to the other family members? 
This could help to not share all private photos between users, just the common (external) one, but does not resolves people match.

**Comment by PrayerfulDrop** [comment]
The lack of sharing my 120k+ assets with my family users automatically has caused me to halt using Immich.  There is no easy way to create albums based on the assets and lack of global select all capabilities.  I do not use folder structures in the external library since 2015 so the CLI tool is useless.  All of my assets from iCloud into years/months/days folder structure.  Without the option to expose the entire external library to all IMMICH users or the ability to create albums from the AI and face recognition groupings.

**Comment by andymarden** [comment]
I have outlined this before - it's clearly a big deal to me and others. It is a data modelling thing:

- a user exists
- a library exists (this may be external or internal)
- a library has one user as the owner
- each library has a number of actions that can be performed on it - add, delete, modify photos, link facial recognition, view map info, view facial recognition etc (this will expend over time), view action permissions, modify action permissions, transfer ownership
- any user can have permissions on each action on each library the owner always has permissions for all actions on a library they own

Unless you think if this in the terms of this (simplified) data model, everything will always be a compromise.

At the moment, the inherent data model (explicit or implicit) is not like this and matching the common use cases suffers as a result). You can patch around the edges, but it will get harder and harder to maintain.

Happy to help as I have about a million years experience doing this sort of stuff and the scars to show.

**Comment by eddified** [comment]
> My setup is this: I have a "family" user which owns an external library which is where all of our family's historical photos are. This is then shared with each family member as a user.

I cannot find a way to share the external library. @andymarden can you please explain how you did that?

**Reply by kgerg2** [reply]
Not OP, but I have a similar setup. There is a user for the sole purpose of being the owner of an external library, and it shares all its photos using the [partner sharing](https://immich.app/docs/features/partner-sharing) feature, essentially sharing the external library. This seems to me the easiest way to do it.

**Reply by hs-ye** [reply]
this should really go to the top as the current best workaround solution, it basically achieves what most of the use cases I imagine people are using Immich for in terms of sharing privately within families.

**Reply by davidancell** [reply]
I use the same setup as @kgerg2 however, the people matching feature isn't available. It's very unfortunate. I'd LOVE for this to be implemented.

**Comment by enlight3d** [comment]
Hello, I have the same use case as others here. Migrating from synology photos to immich and have a shared library that is not tied to a user would be a very nice feature. I think that each user should have their own libraries and this can stay as it is right now, but the external libraries (such as a docker mount point from lets say a shared volume /Photos folder of a NAS) could be made as available to other users according to some permissions.

**Reply by fussel132** [reply]
I am in the exact same situation right now moving to Immich from Synology. This "shared space" is not reproducable in Immich with all the features right? Would be nice to have one (or even multiple) libraries and for pictures/albums you have the option to move them to the shared library or keep in your personal one. Similar to how iCloud Shared Library works...

**Reply by chriswhaas** [reply]
Same here, I want to switch from Synology photos to Immich, but In need one "shared Space" for all members and my private space next to each other. "Shared Space" is for family images, private for job documentation pictures, for example. The backup should always into my private Space first. From there I decide, which images should go into the "shared space" for all. Thank you very much.

**Reply by andymarden** [reply]
Can we please restructure for this? I will help with the data model like I said - it will be worth it.

**Reply by chriswhaas** [reply]
One more thing: I never liked the face recognition system in Synology photos, because they are separated between Shared space and private space, so I had to learn and name each person in both spaces again. It would be great, if each face recognition set would reflection in both worlds, including the name of the person. In my account I should see all Person A images from shared and my private space in one virtual folder of this person. My partner should see in her account all images from Person A in shared space and her private space in her virtual folder of this person A. The Naming of Person A should also be shared as soon as this Person A is available either in shared space or her private space for the first time.

**Reply by andymarden** [reply]
Yes - part of the data model revision I have in mind is to separate out users from libraries and then sharing on photos, facial recognition, maps etc so that each aspect is treated as a data class .

**Reply by 66tree** [reply]
> Same here, I want to switch from Synology photos to Immich, but In need one "shared Space" for all members and my private space next to each other. "Shared Space" is for family images, private for job documentation pictures, for example. The backup should always into my private Space first. From there I decide, which images should go into the "shared space" for all. Thank you very much.

I have scanned thousands of old pictures that I want to do this with as well. Right now they are in an external library that I wish I could share "view only" rights to every new user added to Immich.

**Comment by nghiemlong** [comment]
As a Photo vault for families, the fact that Immich does not allow Location & People sharing via partner link make me confuse. The map is basically useless for me because I need the whole family see it, not only me. 

Sorry I found that a partner can see photo map of shared albums. So the only feature I need now is People sharing.

**Reply by coltenkrauter** [reply]
I’m not sure this project was originally marketed as a photo vault for families, but that’s how I’m using it. I’d definitely appreciate the ability to share faces and location data with my family members.

**Reply by alextran1502** [reply]
Just an update from the team. As we are getting close to finishing all the work for Kysely migration, we are starting to discuss on how to implement this feature request

**Reply by coltenkrauter** [reply]
That is great news. Is there a separate issue that covers this?

**Reply by alextran1502** [reply]
Yes. Although the discussion is done internally

https://github.com/immich-app/immich/issues/12614

**Reply by nghiemlong** [reply]
Sorry I found that a partner can see photo map of shared albums. So the only feature I need now is People sharing.

**Reply by andymarden** [reply]
> Just an update from the team. As we are getting close to finishing all the work for Kysely migration, we are starting to discuss on how to implement this feature request

See above - it's my view that the data model is the starting point of your want to do it properly.

**Comment by srvprivate** [comment]
I too am wanting this. As it being a Family User for others to upload and view to, tagging the people is important, but those people don't show on my OWN user as people. Which means I would need to upload my own library in order to tag them. This hybrid approach is very crucial, and something like "Tagged" should exist (like Facebook started in the early days and still has). A hyper link to each person and allowing that data to be toggled on or off in sharing settings. We need more Share Permissions.

**Reply by MilesTEG1** [reply]
Hello,
This is the thing for what I can't migrate to Immich.
(Plus the fact that admin can't delete photos from another user...)

Hope this will be added soon.

**Reply by chriswhaas** [reply]
same here!

**Reply by larswinkelmann** [reply]
same here! Hope this will be added soon.

**Reply by srvprivate** [reply]
@alextran1502  `Revisiting this after testing with family accounts, the workflow is still pretty cumbersome in practice. Not a knock on the project (totally understand this was originally built for you and your wife and has since blown up), but here’s what feels missing from a family-use perspective:

- **Shared/External Library Model**
    
    - Ability to create a dedicated **Shared Library** that multiple accounts can upload to, instead of relying on one “owner” account.
        
    - Clear **UI section for Shared** with simple toggles/permissions (view, contribute, comment, tag, etc.).
        
    - Upload flow should let users choose “Personal” vs “Shared” library (or set a default).
        
- **Unified People/Face Data**
    
    - Face tagging and people mapping in a Shared Library should propagate to all accounts with access, instead of staying siloed to whoever did the upload.
        
- **Template-Based Accounts (“All-On Family Mode”)**
    
    - When creating new family accounts, it would be huge if they could inherit a **default template** with Partner/Sharing already enabled.
        
    - Basically: avoid deep settings for each person → no need to enable Partner sharing or re-share each album individually.
        
    - One toggle like **“Family Mode”** that automatically applies shared access + default partner settings across all accounts.
        

The core issue: not everyone in a family is tech-savvy. The workflow needs to be as “stupid simple” as possible: sign in, upload, pick where it goes. Done.`

Curious, I know you said above, you're working on it internally, where does that seem to land in progress? Trying to figure out how to navigate with my family moving forward after a funeral and want to not give a bunch of changing directives if I made them separate accounts! Much love bro!

**Reply by amacal1** [reply]
> @alextran1502 `Revisiting this after testing with family accounts, the workflow is still pretty cumbersome in practice. Not a knock on the project (totally understand this was originally built for you and your wife and has since blown up), but here’s what feels missing from a family-use perspective:
> 
>     * **Shared/External Library Model**
>       
>       * Ability to create a dedicated **Shared Library** that multiple accounts can upload to, instead of relying on one “owner” account.
>       * Clear **UI section for Shared** with simple toggles/permissions (view, contribute, comment, tag, etc.).
>       * Upload flow should let users choose “Personal” vs “Shared” library (or set a default).
> 
>     * **Unified People/Face Data**
>       
>       * Face tagging and people mapping in a Shared Library should propagate to all accounts with access, instead of staying siloed to whoever did the upload.
> 
>     * **Template-Based Accounts (“All-On Family Mode”)**
>       
>       * When creating new family accounts, it would be huge if they could inherit a **default template** with Partner/Sharing already enabled.
>       * Basically: avoid deep settings for each person → no need to enable Partner sharing or re-share each album individually.
>       * One toggle like **“Family Mode”** that automatically applies shared access + default partner settings across all accounts.
> 
> 
> The core issue: not everyone in a family is tech-savvy. The workflow needs to be as “stupid simple” as possible: sign in, upload, pick where it goes. Done.`
> 
> Curious, I know you said above, you're working on it internally, where does that seem to land in progress? Trying to figure out how to navigate with my family moving forward after a funeral and want to not give a bunch of changing directives if I made them separate accounts! Much love bro!

This is the most succinct, yet complete, definition of the missing features and - I think - what we all want to see. It's perfect. I care more about the "Unified Face / People Data" part of it, but all of this would be useful and nice to have. 

For me, personally, I desperately want to see face/object definition shared between users. I don't necessarily care to always search through my wife's photos for my kid's faces, but occasionally might be nice. What I DO want is for either my wife or I to be able to search the massive external libraries that are shared between us. She has her photos, I have my photos, and - lately - even my daugther is using Immich on her devices. But we also have a decade of "family" photos that we would all like to have the ability to see (currently works), locate on a map (currently works), and find ourselves in them (currently does not work), or find our pets/bikes/clouds, etc. in them (currently does not work).

**Comment by andymarden** [comment]
I would do a pull request in this but the change required is reasonably extensive in changing the underlying data model. I have offered to help in that, but it seems to fall on deaf ears.

**Comment by thiagomg** [comment]
I have a similar use case and I´d love to see this implemented. Being:
Me, my wife and my daughters each have our phones, from which I´d love to have the pictures taken going to a each library, but always shared to each other. Secondly, both me and one of my daughters have digital cameras and the pictures taken from this cameras go to a separate location as I need to manually copy the photos. I can just create a separate user for the camera photos, but as we'd have this shared location, I'd love to see all my wife pictures, for instance, when I select her face, regardless of the user owning that photo.
That said, I am a software developer myself and I could help building this feature, so it would be great to:
1. Know whether this is something that you are planning on implementing
2. Have some pointers where this needs to be implemented, so I can create pull requests.

Thanks!

**Reply by sdf121234** [reply]
I have the same problem. I absolutely need face recognition for shared photos. I would already be satisfied if the faces had to be retrained and could not be reused from personal photos. For me personally, it would be important that other users have access to the face database of the shared photos. Unfortunately, until this is possible, immich is not usable for me yet. :-(

**Comment by txemi** [comment]
This discussion describes a very rich vision of shared libraries + shared people matching across users.

In case it’s useful to have a smaller, easier first step on the roadmap, I’ve opened a narrow‑scope feature request that only asks for local face recognition on already shared assets, without sharing any ML/People data between accounts:

https://github.com/immich-app/immich/discussions/24290

It’s meant as a pragmatic building block that could ship earlier, while bigger “shared ML” ideas mature.

**Comment by andymarden** [comment]
Great thanks.
  

  

  

  
  
  
  
  
>

**Comment by Vinticb** [comment]
**Expanded Proposal: Collaborative People Management with Suggestion System**
I'd like to expand on this excellent feature request with a more detailed collaborative approach that could greatly enhance the family sharing experience.

**Three-Tier People Management Architecture**
1. Shared People (Family Library)
- Admin-curated central database - One or more administrators manage the core family library
- Automatic synchronization - Shared people automatically appear in all authorized family members' accounts
- Read-only for regular users - Family members can use shared identifications but cannot modify them directly
- Consistent metadata - Names, birth dates, and profile pictures sync across all users

2. Private People (Individual Library)
- Personal identifications - Each user maintains their own private people database
- Complete privacy - Private people visible only to the creating user
- Full editing control - Users can freely add, edit, or delete their private people

3. Collaborative Suggestion System
- "Suggest for Family Library" - Any family member can propose moving their private people to the shared library
- "Improve this Person" - Users can suggest better names, profile pictures, or metadata for existing shared people
- Admin approval workflow - All suggestions require administrator review and approval
- Notification system - Admins get notified of pending suggestions, users get status updates

**Example Workflows**
Adding New Shared Person:
_User identifies "Cousin Alex" as private person 
→ Clicks "Suggest for Family Library" 
→ Adds reason: "This is our cousin who visits often"
→ Admin receives notification, reviews suggestion  
→ Admin approves → "Cousin Alex" moves to shared library for all users_

**Improving Existing Person:**
_User views shared person "Grandma Mary"
→ Notices profile picture is blurry
→ Clicks "Suggest Better Photo", selects clearer image
→ Admin reviews and approves better photo
→ All users now see updated profile picture_

**Key Benefits Over Simple Sharing**
- Collaborative Curation - Everyone can contribute to family library quality
- Reduced Admin Burden - Distributed identification workload across family members
- Quality Control - Admin maintains final approval authority
- Privacy Respect - Users keep personal contacts private while sharing family members
- Continuous Improvement - Library evolves with collective family knowledge

**Implementation Considerations**

- Opt-in participation - Users choose when to make suggestions
- Clear visual indicators - Distinguish between shared vs private people
- Suggestion history - Track what changes were proposed and by whom
- Bulk operations - Admins can process multiple suggestions efficiently

This approach maintains the core sharing concept discussed here but adds much-needed collaborative curation capabilities that reflect how families actually work together on photo management.

The suggestion system could be a fantastic middle ground between completely siloed people databases and fully open editing - giving everyone a voice while keeping final control where it belongs for family harmony! 🏠✨

**Reply by 66tree** [reply]
In my opinion this would be _the_ 3.0 headline feature.

**Reply by chriswhaas** [reply]
Sounds very good for me!

**Reply by TKaluza** [reply]
Yes but " → Admin reviews and approves better photo in Improving Existing Person" 
I trust my family, so just a history to take back the last changes would be great.

**Comment by drewmerc302** [comment]
I just spent 45 minutes trying to figure out why my wife couldn't see my external library on her user account even after i enabled partner sharing. I just stumbled across this thread. For me, Immich is essentially worthless without this feature and i certainly wouldn't have spent the better part of a week installing, configuring, and letting import jobs run on 200,000+ photos/videos if i had known the only person that could see them would be me.

+1 on the ability to share an existing external library with family

**Reply by sneumann** [reply]
> I just spent 45 minutes trying to figure out why my wife couldn't see my external library on her user account even after i enabled partner sharing. I just stumbled across this thread. For me, Immich is essentially worthless ...

Calm down please, you are talking to volunteers and not a company that sold you something for $$$. 
I have same needs here, but I might have a solution:

> +1 on the ability to share an existing external library with family
With some googl'ing and help from a friendly LLM (so careful, it might eat kitten!) I created 
https://gist.github.com/sneumann/dbb1de943cd28fb52c8411f027e063b2

Change the URL to your immich server. 
This will search for all assets that are `"isExternal": True`, so best if you have only one external library.
It will then create an album `ALBUM_NAME = "External Library Name"`, and you can make it match the name of the external library. Use `--dry-run` first to get a feeling, check the number of assets it wants to include in the Album.
Execution was blazing fast for 16K images, so can't take too long for 200K. I only needed to run this once, but during my search I came across https://github.com/Salvoxia/immich-folder-album-creator which can even create multiple albums from a folder hierarchy. Once your external library is in Albums, that can be shared. 
Looking forward to see a solution to the people/face recognition sharing.

Yours,
Steffen

**Reply by arnschi** [reply]
>  For me, Immich is essentially worthless without this feature and i certainly wouldn't have spent the better part of a week installing, configuring, and letting import jobs run on 200,000+ photos/videos if i had known the only person that could see them would be me.

That's why you try things out in a small playground, before going big on your full life.

Yes, I am also waiting for a "real" solution for people sharing. Until now, if you don't like it, you can for sure get the money back you paid for this tool.

**Reply by drewmerc302** [reply]
Thanks all. i've come to terms with the limitations and, based on the price of entry, i'm fully satisfied with what i can do. Anything and everything is better than paying for hosting somewhere else. all the help is much appreciated

**Comment by sneumann** [comment]
Hi, while a real solution takes some time, is there any way for a hackish solution ? 
Could one dump all people and their location in assets/images through the API by one user, 
and import by another user ? That would "only" be, say, a (few?) kilobyte of JSON per person duplicated. 
It would not syncronised, so you'd have a user that does all the annotations, but all others who imported that would see and use the people annotations. Yours, Steffen

**Comment by photoserver-pro** [comment]
@andymarden Your data model framing resonates strongly. I'd like to add a complementary UX perspective.

In many real-world setups — especially families consolidating years of photos from multiple devices, NAS systems, and laptops — the natural unit of sharing is not the album, not the individual asset, but the external library itself. Each family member has their own storage, their own collection, and they want to bring it into a shared system without giving up ownership.

This shifts the permission question from "how do I share my photos?" to "which users should have access to which library?" — and that's a much cleaner model to reason about.

I've described this as a concrete user story in https://github.com/immich-app/immich/issues/12614:

4 external libraries, each belonging to a different family member
7 users on the same instance
User A has access to one library, User B to two, User C to all four
Metadata — people, tags, GPS — follows library access automatically
The admin's role is purely operational: managing the system, not owning most of the content. That's a meaningful distinction from the current model.

For broader context on navigation and discovery in multi-user archives, I've also written a longer piece here: https://github.com/immich-app/immich/discussions/27151

There might also be an interesting model of how external library sharing could look in practice — I'm currently looking at my staging envirorment at this: https://github.com/open-noodle/gallery/discussions/133 . @Deeds67 has a very modern approach that works well.

**Comment by andymarden** [comment]
Two exactly - that framing:   "how do I share my photos?" to "which users should have access to which library?" - first assumed user is a data entity and that it has a child entity of photos (or library); the second has two entities: user and library.   
  

  
That is at the heart of the issue. It's the kind of thing that creeps in when the data model is not foundational (or, to be fair, where it is and the use cases have not been broad enough).
  

  

  

  
  
  
  
>

**Comment by felixfuchs90** [comment]
We use Immich on Unraid for our extended family. Multiple users upload photos of the same people (children, grandparents, family events). Shared face recognition and shared people management would be one of the most valuable features for us.

**Comment by svaens** [comment]
Firstly, congratulations to the developers - Immich looks really nice, and on the whole, works well. 

I setup Immich recently, with certain assumptions in mind (stupid me) about how I was going to be able to use it.
I was shocked and dismayed when, my wife couldn't (even as 'partner' in the app) add photos I've imported to her 'favorites' list. 

OK, never mind. I have thought of a workaround. Simply create an album, and add them to that. 
**Denied**. 

Frankly, I can't help wondering if some of my current problems are not so much 'design issues' but 'bug', because, she is able to add photos from the context of the album, but she can't add them when just browsing/filtering photos.  Probably the feature of 'Partner' needs to be discarded, and replaced with a more universal permissions schema. A bit like the Linux/Unix file-system, with Uses/Groups/Others, and a whole bunch of roles (instead of just read,write,execute). 

But these are all wishes of what I need, as the administrator of an Immich server, which I set up for my family. As it is, it is not so useful (for the family). But I hope these use-cases are considered in the 'soon to be implemented' improvements (as listed in the Roadmap as "Better sharing" and "User Groups"). 

Anyway, thanks again to the developers for developing such a nice piece of software, and making it so readily available to the masses.

**Comment by ag3ntfury** [comment]
I would like to suggest another aspect for family libraries.
In a self-hosted family setup, the server administrator should manage the infrastructure (users, backups, updates, storage), but should not automatically have access to every user's photos.
By default, each user's library would remain private.
If a user needs help, they could explicitly delegate access to the administrator (or another trusted user) with selectable permissions such as View, Edit, or Delete, and optionally set an expiration time. The user could revoke access at any time.
This would allow administrators to help family members without permanently having unrestricted access to their private libraries.
